File: 095_UniqueBinarySearchTreesII.cpp
// Topic: Unique Binary Search Trees II
// Description: Generate all structurally unique BSTs that store values 1..n.
// Approach: Recursively generate trees by choosing each root and combining left/right subtrees.

#include <bits/stdc++.h>
using namespace std;

struct TreeNode
{
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};

vector<TreeNode*> generateTreesRange(int lo, int hi)
{
    vector<TreeNode*> res;
    if (lo > hi)
    {
        res.push_back(nullptr);
        return res;
    }
    for (int root = lo; root <= hi; ++root)
    {
        auto lefts = generateTreesRange(lo, root - 1);
        auto rights = generateTreesRange(root + 1, hi);
        for (auto L : lefts)
            for (auto R : rights)
            {
                TreeNode* node = new TreeNode(root);
                node->left = L;
                node->right = R;
                res.push_back(node);
            }
    }
    return res;
}

vector<TreeNode*> generateTrees(int n)
{
    if (n == 0)
        return {};
    return generateTreesRange(1, n);
}

// helper: preorder serialize with null markers
void serializePreorder(TreeNode* root, string& out)
{
    if (!root)
    {
        out += "# ";
        return;
    }
    out += to_string(root->val) + " ";
    serializePreorder(root->left, out);
    serializePreorder(root->right, out);
}

int main()
{
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    vector<int> tests = {2, 3};
    for (int i = 0; i < (int)tests.size(); ++i)
    {
        auto trees = generateTrees(tests[i]);
        cout << "Test " << (i + 1) << " (n=" << tests[i] << "): count=" << trees.size() << '\n';
        for (auto t : trees)
        {
            string s;
            serializePreorder(t, s);
            cout << s << "\n";
        }
    }
    return 0;
}

/*
Time Complexity: Catalan number growth — superexponential in n (O(C_n) trees), each tree
construction costs O(n) Space Complexity: O(C_n * n) for result
*/

File: 095_UniqueBinarySearchTreesII.py
# Topic: Unique Binary Search Trees II
# Description: Generate all structurally unique BSTs that store values 1..n.
# Approach: Recursively generate trees by choosing each root and combining left/right subtrees.

from typing import Optional, List


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def generateTrees(self, n: int) -> List[Optional[TreeNode]]:
        if n == 0:
            return []

        def generate(lo, hi):
            if lo > hi:
                return [None]
            trees = []
            for root_val in range(lo, hi + 1):
                lefts = generate(lo, root_val - 1)
                rights = generate(root_val + 1, hi)
                for L in lefts:
                    for R in rights:
                        root = TreeNode(root_val)
                        root.left = L
                        root.right = R
                        trees.append(root)
            return trees

        return generate(1, n)


def serialize_preorder(root):
    if not root:
        return "#"
    return f"{root.val} {serialize_preorder(root.left)} {serialize_preorder(root.right)}"


if __name__ == "__main__":
    tests = [2, 3]
    for i, n in enumerate(tests):
        trees = Solution().generateTrees(n)
        print(f"Test {i + 1} (n={n}): count={len(trees)}")
        for t in trees:
            print(f"  {serialize_preorder(t)}")

"""
Time Complexity: O(4^n / sqrt(n)) Catalan number
Space Complexity: O(4^n / sqrt(n)) for storing all trees
"""
